compiler/rustc_trait_selection/src/solve/opaques.rs

```diff
@@ -0,0 +1,62 @@
+use rustc_middle::traits::query::NoSolution;
+use rustc_middle::traits::solve::{Certainty, Goal, QueryResult};
+use rustc_middle::traits::Reveal;
+use rustc_middle::ty;
+use rustc_middle::ty::util::NotUniqueParam;
+
+use super::{EvalCtxt, SolverMode};
+
+impl<'tcx> EvalCtxt<'_, 'tcx> {
+    pub(super) fn normalize_opaque_type(
+        &mut self,
+        goal: Goal<'tcx, ty::ProjectionPredicate<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        let tcx = self.tcx();
+        let opaque_ty = goal.predicate.projection_ty;
+        let expected = goal.predicate.term.ty().expect("no such thing as an opaque const");
+
+        match goal.param_env.reveal() {
+            Reveal::UserFacing => match self.solver_mode() {
+                SolverMode::Normal => {
+                    if !self.can_define_opaque_ty(opaque_ty.def_id) {
+                        return Err(NoSolution);
+                    }
```

Comment:
at some point we should call queries without defining new opaque types but having the existing opaque type definitions.

for this we have to move that down to until after we've checked the existing opaque types

Reply:
I'll leave a FIXME that we should experiment with this, but I'm not gonna mess with this.